### PR TITLE
Add a possibility to push and pull files from both simulator and real devices

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Added a possibility to push/pull files to/from real devices via SSHD file system (osxfuse/ifuse combination). Also, it is now possible to access files inside application containers.

### CODE REVIEW

File: lib/commands/actions.js [hunk 1]
```diff
@@ -1,23 +1,231 @@
...
+ *                              to the default media folder and only the file name is considered important.
+ * @param {string} base64Data - Base-64 encoded content of the file to be uploaded.
+ */
+async function pushFileToSimulator (device, remotePath, base64Data) {
+  if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
+    if (!remotePath.includes('/')) {
+      log.errorAndThrow(`It is expected that bundle identifier is separated from the relative path with a single slash. ` +
+                        `'${remotePath}' is given instead`);
+    }
+    const bundleId = remotePath.substring(CONTAINER_PATH_MARKER.length, remotePath.indexOf('/'));
```
Comment (imurchie) [inline lib/commands/actions.js]: All of this seems built for a regular expression 
```js
let remotePath = '@com.myapp.bla/RelativePathInContainer/111.png';
if (if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
  let match = /@([^\/]+)\/(.+)/.exec(remotePath)
  if (match) {
    let bundleId = match[1];
    let relativePath = match[2];
    // ...
  } else {
    // throw error
  }
}
```

File: lib/commands/actions.js [hunk 2]
```diff
@@ -1,23 +1,231 @@
...
+ * @return {string} Base-64 encoded content of the remote file
+ */
+async function pullFileFromRealDevice (device, remotePath) {
+  await verifyIFusePresence();
+  const mntRoot = await tempDir.tempDir();
+  let isUnmountSuccessful = true;
+  try {
+    let dstPath = path.resolve(mntRoot, remotePath);
+    let ifuseArgs = ['-u', device.udid, mntRoot];
+    if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
```
Comment (imurchie) [inline lib/commands/actions.js]: This is beginning to be repetitive. The code for getting the bundleId and relative path could be extracted out.

File: lib/commands/actions.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/commands/actions.js]: fixed

File: lib/commands/actions.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/commands/actions.js]: extracted

File: lib/commands/actions.js [hunk 3]
```diff
@@ -1,23 +1,232 @@
...
+ *                              The relative folder path is ignored if the file is going to be uploaded
+ *                              to the default media folder and only the file name is considered important.
+ * @param {string} base64Data - Base-64 encoded content of the file to be uploaded.
+ */
+async function pushFileToSimulator (device, remotePath, base64Data) {
+  if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
+    const [bundleId, dstPath] = await parseContainerPath(remotePath,
+      async (x) => await getAppContainer(device.udid, x));
+    log.info(`Parsed bundle identifier '${bundleId}' from '${remotePath}'. ` +
+             `Will put the data info '${dstPath}'`);
```
Comment (imurchie) [inline lib/commands/actions.js]: Typo: `info` -> `into`

File: lib/commands/actions.js [hunk 4]
```diff
@@ -1,23 +1,232 @@
...
+  const mntRoot = await tempDir.tempDir();
+  let isUnmountSuccessful = true;
+  try {
+    let dstPath = path.resolve(mntRoot, remotePath);
+    let ifuseArgs = ['-u', device.udid, mntRoot];
+    if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
+      const [bundleId, pathInContainer] = await parseContainerPath(remotePath, mntRoot);
+      dstPath = pathInContainer;
+      log.info(`Parsed bundle identifier '${bundleId}' from '${remotePath}'. ` +
+               `Will put the data info '${dstPath}'`);
```
Comment (imurchie) [inline lib/commands/actions.js]: Typo.

File: lib/commands/actions.js [hunk 5]
```diff
@@ -1,23 +1,232 @@
...
+      ifuseArgs = ['-u', device.udid, '--container', bundleId, mntRoot];
+    }
+    await mountDevice(device, ifuseArgs);
+    isUnmountSuccessful = false;
```
Comment (imurchie) [inline lib/commands/actions.js]: Why not just start with `isUnmountSuccessful` in the `false` state and set it to `true` after unmount.

Comment (mykola-mokhnach) [reply lib/commands/actions.js]: because I still want to remove that temporary folder even if mount fails

File: lib/commands/actions.js (hunk 4 above)
Comment (mykola-mokhnach) [reply lib/commands/actions.js]: fixed

File: lib/commands/actions.js (hunk 3 above)
Comment (mykola-mokhnach) [reply lib/commands/actions.js]: fixed

**Pull request merged**